File: tests/orig_parity/02_mpdta_original.py
"""StatsPAI original-data parity (Python side) -- Module 02.

Runs sp.callaway_santanna on the *original* did::mpdta extract.
"""

from __future__ import annotations

import statspai as sp

from _common import OrigRecord, read_csv, write_results

MODULE = "02_mpdta_original"


def main() -> None:
    df = read_csv(MODULE)
    n = len(df)

    fit = sp.callaway_santanna(
        df,
        y="lemp",
        g="first_treat",
        t="year",
        i="countyreal",
        estimator="reg",
        control_group="nevertreated",
    )
    dyn = sp.aggte(fit, type="dynamic")

    # The did vignette does not print a simple-aggregation ATT for
    # mpdta, so the simple row carries no published anchor: the claim
    # is same-byte parity against did::aggte(type="simple").  The
    # vignette *does* print the overall dynamic/event-study ATT
    # (-0.0772, xformla = ~1), which anchors the second row.
    rows = [
        OrigRecord(
            module=MODULE,
            side="py",
            statistic="simple_ATT",
            estimate=float(fit.estimate),
            se=float(fit.se),
            n=n,
            published=None,
            citation="did::aggte(type='simple') on the same bytes; no printed vignette anchor",
        ),
        OrigRecord(
            module=MODULE,
            side="py",
            statistic="dynamic_overall_ATT",
            estimate=float(dyn.estimate),
            se=float(dyn.se),
            n=n,
            published=-0.0772,
            citation="Callaway-Sant'Anna 'did' vignette (did-basics), overall dynamic ATT, xformla=~1",
        ),
    ]

    write_results(MODULE, "py", rows, extra={"data_source": "did::mpdta", "n_obs": n})


if __name__ == "__main__":
    main()

File: tests/orig_parity/02_mpdta_original.R
# Original-data parity for did::mpdta (the Callaway-Sant'Anna 2021
# vignette panel). The did vignette does not print a simple-aggregation
# ATT for mpdta, so that row carries no published anchor; the vignette's
# printed overall dynamic/event-study ATT (-0.0772, xformla = ~1)
# anchors the second row.

.script_dir <- (function() {
  args <- commandArgs(trailingOnly = FALSE)
  m <- grep("^--file=", args, value = TRUE)
  if (length(m) > 0) dirname(normalizePath(sub("^--file=", "", m[1])))
  else getwd()
})()
DATA_DIR <- file.path(.script_dir, "data")
RESULTS_DIR <- file.path(.script_dir, "results")
dir.create(DATA_DIR, showWarnings = FALSE, recursive = TRUE)
dir.create(RESULTS_DIR, showWarnings = FALSE, recursive = TRUE)

suppressPackageStartupMessages({
  library(did)
  library(jsonlite)
})

data("mpdta", package = "did")

# Rename to the StatsPAI convention so the Python side can read it
# without a column-renaming step.
out <- mpdta
names(out)[names(out) == "first.treat"] <- "first_treat"
write.csv(out, file.path(DATA_DIR, "02_mpdta_original.csv"),
          row.names = FALSE)

# Run R-side did::att_gt + aggte simple ATT.
fit <- did::att_gt(
  yname = "lemp", tname = "year", idname = "countyreal",
  gname = "first.treat", data = mpdta,
  control_group = "nevertreated",
  est_method = "reg", bstrap = FALSE
)
agg <- did::aggte(fit, type = "simple", bstrap = FALSE, cband = FALSE)
dyn <- did::aggte(fit, type = "dynamic", bstrap = FALSE, cband = FALSE)

build_row <- function(stat, est, se, n, published, citation) {
  list(module = jsonlite::unbox("02_mpdta_original"),
       side = jsonlite::unbox("R"),
       statistic = jsonlite::unbox(stat),
       estimate = jsonlite::unbox(est),
       se = jsonlite::unbox(se),
       n = jsonlite::unbox(n),
       published = jsonlite::unbox(published),
       citation = jsonlite::unbox(citation),
       extra = list())
}

rows <- list(
  build_row("simple_ATT",
            agg$overall.att, agg$overall.se,
            nrow(mpdta),
            NA,
            "did::aggte(type='simple') on the same bytes; no printed vignette anchor"),
  build_row("dynamic_overall_ATT",
            dyn$overall.att, dyn$overall.se,
            nrow(mpdta),
            -0.0772,
            "Callaway-Sant'Anna 'did' vignette (did-basics), overall dynamic ATT, xformla=~1")
)

payload <- list(
  module = jsonlite::unbox("02_mpdta_original"),
  side = jsonlite::unbox("R"),
  rows = rows,
  extra = list(data_source = jsonlite::unbox("did::mpdta"),
               n_obs = jsonlite::unbox(nrow(mpdta)))
)
writeLines(
  jsonlite::toJSON(payload, pretty = TRUE, na = "null", null = "null",
                    digits = NA),
  file.path(RESULTS_DIR, "02_mpdta_original_R.json")
)
message("OK -- wrote 02_mpdta_original_R.json")
